src/UnityExtension/Assets/Editor/GitHub.Unity/UI/TreeControl.cs

```diff
@@ -257,26 +255,116 @@ public void Blur()
+        {
+            for (var i = idx + 1; i < nodes.Count && node.Level <= nodes[i].Level; i++)
             {
-                switch (node.CheckState)
+                var childNode = nodes[i];
+                if (node.Level == childNode.Level)
                 {
-                    case CheckState.Empty:
-                        node.CheckState = CheckState.Checked;
-                        break;
+                    continue;
+                }
+
+                childNode.CheckState = isChecked ? CheckState.Checked : CheckState.Empty;
 
-                    case CheckState.Checked:
-                        node.CheckState = CheckState.Empty;
-                        break;
+                if (childNode.IsFolder)
+                {
+                    ToggleChildrenChecked(i, childNode, isChecked);
                 }
+            }
+        }
+
+        private void ToggleParentFoldersChecked(int idx, TreeNode node, bool isChecked)
+        {
+            while (true)
+            {
+                if (node.Level > 0)
+                {
+                    var siblingsInSameState = true;
+                    var firstSiblingIndex = idx;
+
+                    for (var i = idx - 1; i > 0 && node.Level <= nodes[i].Level; i--)
+                    {
+                        var previousNode = nodes[i];
+                        if (node.Level < previousNode.Level)
+                        {
+                            continue;
+                        }
+
+                        firstSiblingIndex = i;
+
+                        if (siblingsInSameState)
+                        {
+                            var previousNodeIsChecked = previousNode.CheckState == CheckState.Checked;
 
-                Debug.LogFormat("Ripple CheckState index:{0} level:{1}", idx, node.Level);
+                            if (isChecked != previousNodeIsChecked)
+                            {
+                                siblingsInSameState = false;
+                            }
+                        }
+                    }
+
+                    if (siblingsInSameState)
+                    {
+                        for (var i = idx + 1; i < nodes.Count && node.Level <= nodes[i].Level; i++)
+                        {
+                            var followingNode = nodes[i];
+                            if (node.Level < followingNode.Level)
+                            {
+                                continue;
+                            }
+
+                            var followingNodeIsChecked = followingNode.CheckState == CheckState.Checked;
+                            if (isChecked != followingNodeIsChecked)
+                            {
+                                siblingsInSameState = false;
+                                break;
+                            }
+                        }
+                    }
+
+                    var parentIndex = firstSiblingIndex - 1;
```

Comment:
Same here for firstSiblingIndex not being 0?

Reply:
We would not have entered this routine if the level was not at least 1.
So the first index this will ever work on is 1. Meaning our we can safely expect to find a parent node at 0.